libraries/chain/block_state.cpp

```diff
@@ -102,5 +102,45 @@ namespace eosio::chain {
    ,_cached_trxs( std::move(trx_metas) )
    {}
 #endif
+
+   std::optional<qc_data_t> block_state::get_best_qc() const {
+      auto block_number = block_num();
+
+      // if pending_qc does not have a valid QC, consider valid_qc only
+      if( !pending_qc.is_valid() ) {
+         if( valid_qc ) {
+            return qc_data_t{ quorum_certificate{ block_number, valid_qc.value() },
+                              qc_info_t{ block_number, valid_qc.value().is_strong() }};
+         } else {
+            return std::nullopt;;
+         }
+      }
+
+      // extract valid QC from pending_qc
+      valid_quorum_certificate valid_qc_from_pending(pending_qc);
+
+      // if valid_qc does not have value, consider valid_qc_from_pending only
+      if( !valid_qc ) {
+         return qc_data_t{ quorum_certificate{ block_number, valid_qc_from_pending },
+                           qc_info_t{ block_number, valid_qc_from_pending.is_strong() }};
+      }
+
+      // Both valid_qc and valid_qc_from_pending have value. Compare them and select a better one.
+      // Strong beats weak. Break tie with highest accumulated weight.
+      auto valid_qc_is_better = false;
+      if( valid_qc.value().is_strong() && !valid_qc_from_pending.is_strong() ) {
+         valid_qc_is_better = true;
+      } else if( !valid_qc.value().is_strong() && valid_qc_from_pending.is_strong() ) {
+         valid_qc_is_better = false;
+      } else if( valid_qc.value().accumulated_weight() >= valid_qc_from_pending.accumulated_weight() ) {
+         valid_qc_is_better = true;
+      } else {
+         valid_qc_is_better = false;
```

Comment:
For modularity, I believe it would be better to have a comparison operator in `valid_quorum_certificate`, such as:

```c++
      bool operator<(const valid_quorum_certificate& o) {
         if (is_strong() != o.is_strong())
            return o.is_strong();
         if (accumulated_weight() != o.accumulated_weight()) 
            return accumulated_weight() <  o.accumulated_weight();
         return false;
      }
```

Reply:
Thanks!